### PR TITLE
Feature: ExtensionPoint PrototypeDetailsFactory

### PR DESCRIPTION
**Pull request by mhsdesign**
enable it to override it like:

```yaml
Sitegeist\Monocle\Domain\PrototypeDetails\PrototypeDetailsFactoryInterface:
  className: Foo\Bar\MyFactory

```

Extension point for https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/pull/1

...

Description is WIP will add some more details later

### CODE REVIEW

File: Classes/Domain/PrototypeDetails/UseCases/UseCase.php [hunk 1]
```diff
@@ -14,6 +14,7 @@
  */
 
 use Neos\Flow\Annotations as Flow;
+use Sitegeist\Monocle\Domain\PrototypeDetails\Props\PropValue;
```
Comment (mhsdesign) [inline Classes/Domain/PrototypeDetails/UseCases/UseCase.php:17]: was not imported - but used in doc comment

File: Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php [hunk 2]
```diff
@@ -53,6 +54,11 @@ public function getAnatomy(): Anatomy;
      */
     public function getProps(): PropsCollectionInterface;
 
+    /**
+     * @return UseCaseCollection
+     */
+    public function getUseCases(): UseCaseCollection;
```
Comment (mhsdesign) [inline Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php:60]: missing in interface

File: Classes/Domain/PrototypeDetails/Props/PropValue.php [hunk 3]
```diff
@@ -34,9 +34,12 @@ private function __construct($value)
     {
         if (!self::isValid($value)) {
             throw new \UnexpectedValueException(
-                'PropValue must be a primitive, an array or an object ' .
-                'of type \\stdClass. Got "%s" instead.',
-                is_object($value) ? get_class($value) : gettype($value)
+                sprintf(
```
Comment (mhsdesign) [inline Classes/Domain/PrototypeDetails/Props/PropValue.php:37]: missing sprintf
`is_object($value) ? get_class($value) : gettype($value)` was set as timestamp

**Review by mficzel — CHANGES_REQUESTED** [review]
I do not like that the prototype details factory would be overwritten via objects.yaml for all prototypes. This will not allow to have the styleguide render objects that do not use presentationobjects.

**Comment by mficzel** [comment]
What speaks against configuring this on a per prototype base like many other things. Would be really flexible and still allow to set a default for all presentationObjects.

**Comment by mhsdesign** [comment]
Per prototype basis is to much work (for me as an integrator ^^)

and while per prototype namespace would be working - it would not allow amixed approach per package (which is possible with presentation objects)

i do agree that my proposed extension is not optimal, but it works and is consisting with the currently available monocle extensionpoint see the other service interface for providing the editor (i forgot the name)

**Comment by mhsdesign** [comment]
in a later point we also could make the fusion rendering easier/better extensible.

current problem beeing that only one package can extend monocle per time - for advanced things we need a chain (maybe even provide it as optional Monocle.ExtensionChain package ;) )

if you like we can have a call about this soon ^^

**Comment by mficzel** [comment]
> Per prototype basis is to much work (for me as an integrator ^^)

You do not configure this per prototype as it is in the PresentatioObject:Component predefined

**Comment by mhsdesign** [comment]
aaah you mean because my prototype extends a certain `PackageFactory:PresentationObjectComponent`?

**Comment by mficzel** [comment]
> aaah you mean because my prototype extends a certain PackageFactory:PresentationObjectComponent?

Yes in there you can configure the detailsFactory for presentationobjects and no magic is needed at all. Especially no graceful sidestepping.

**Comment by mhsdesign** [comment]
okay i am starting to get comfortable with the idea ^^

theoretical contra would be:
- if you need another way to determine to use a certain defails factory (instead via the path `@styleguild.@detailsfactory` but via custom logic) -> but can be neglected as we are building this extensionpoint for now only for  https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects

theoretical pro argument:
- one can install monocle with multiple packages providing their own defailsfactory and using all side by side

**Comment by mhsdesign** [comment]
fyi some links i missed previously:



----------

> its planned that when one overrides the interface, it is ones responsibility to also inject the default details service and call it occasionally (when one cannot handle it to fallback to the default)

see fallback logic

https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/blob/040e85037803f9808dcc51870661e38d68692478/Classes/Infrastructure/PrototypeDetailsFactory.php#L69-L72

--------

> i do agree that my proposed extension is not optimal, but it works and is consisting with the currently available monocle extensionpoint see the other service interface for providing the editor (i forgot the name)

i found the name ^^ it is the `PropsCollectionFactoryInterface`
https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/blob/040e85037803f9808dcc51870661e38d68692478/Classes/Infrastructure/PropsCollectionFactory.php#L41-L66

which also uses the fallback logic to moncole

```php
return $this->defaultPropsCollectionFactory->fromPrototypeForPrototypeDetails($prototype);
```

------------

> in a later point we also could make the fusion rendering easier/better extensible.

see how i did it currently ^^ (one way street - only allows either moncle default or the custom presentation/rendering)
https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/blob/040e85037803f9808dcc51870661e38d68692478/Resources/Private/Fusion/Overrides/MonoclePreviewPrototype.fusion#L3-L5

**Comment by mhsdesign** [comment]
by doing this writeup i remebered another contra argument https://github.com/sitegeist/Sitegeist.Monocle/pull/190#issuecomment-1281181327

the proposed interface extraction is only a unplanned extensibility point - since it could have been done purely architectually.

but by introducing something as official as `@styleguild.@detailsfactory` we make lots of monocles internal domain model even more api. As one needs to reuse many services from monocle:

see here: https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/blob/040e85037803f9808dcc51870661e38d68692478/Classes/Infrastructure/PrototypeDetailsFactory.php#L76-L92 only the last line is really different from the default - the rest is copy paste.

my argument is: dont open up to much - there is only this one package in demand for now - and if the time comes craft a nice extensibility api which can controll everything: (tabs at the bottom / rendering / usecase providing / editor for props providing / etc...)

**Comment by mficzel** [comment]
Adding an unplanned extension point with a fixed use case in place is a bit weird. I rather would like to have a clean api for that.

It would be a no go for me to use presentationobjects at all if it would override monocle internals for all prototypes even if it would promise to behave well. If there were a proper api for that I might use po in places.

**Comment by mhsdesign** [comment]
fyi i just found out that using `Sitegeist.Monocle.PropTypes` and `Sitegeist.Monocle.PresentationObjects` side by side cannot work (i assume) as they both override the `PropsCollectionFactoryInterface` (not matter of this pr - but about extensibility and promises ^^)

https://github.com/sitegeist/Sitegeist.Monocle.PropTypes/blob/84a735eaa646d3214afea657c176a32a76d64a0c/Classes/PropsCollectionFactory.php#L22

> > i do agree that my proposed extension is not optimal, but it works and is consisting with the currently available monocle extensionpoint see the other service interface for providing the editor (i forgot the name)
> 
> i found the name ^^ it is the `PropsCollectionFactoryInterface`
> https://github.com/sitegeist/Sitegeist.Monocle.PresentationObjects/blob/040e85037803f9808dcc51870661e38d68692478/Classes/Infrastructure/PropsCollectionFactory.php#L41-L66
> 
> which also uses the fallback logic to moncole
> 
> ```php
> return $this->defaultPropsCollectionFactory->fromPrototypeForPrototypeDetails($prototype);
> ```

this extension point was introduced with https://github.com/sitegeist/Sitegeist.Monocle/releases/tag/v7.3.0

**Comment by mficzel** [comment]
> fyi i just found out that using `Sitegeist.Monocle.PropTypes` and `Sitegeist.Monocle.PresentationObjects` side by side cannot work (i assume) as they both override the `PropsCollectionFactoryInterface` (not matter of this pr - but about extensibility and promises

hmmm yeah would prefer a configuration there aswell ... however one could also say that in that case only adding the interface would be ok. If we already use this pattern we aswell use it twice

File: Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php (hunk 2 above)
Comment (mficzel) [reply Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php:60]: Only if there is a need to add another collection implementation

**Review by mficzel — APPROVED** [review]

File: Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php (hunk 2 above)
Comment (mhsdesign) [reply Classes/Domain/PrototypeDetails/PrototypeDetailsInterface.php:60]: my phpstorm was just confused sometimes ^^

**Pull request merged**